src/dataflow/src/render/mod.rs

```diff
@@ -822,3 +826,334 @@ pub mod datum_vec {
+        },
+        ArrangeBy {
+            input: Box<Plan>,
+            keys: Vec<Vec<MirScalarExpr>>,
+        },
+    }
+
+    impl Plan {
+        /// This method converts a MirRelationExpr into a plan that can be directly rendered.
+        ///
+        /// The rough structure is that we repeatedly extract map/filter/project operators
+        /// from each expression we see, bundle them up
+        pub fn from_mir(expr: &MirRelationExpr) -> Result<Self, ()> {
+            // The plan is to repeatedly attempt to convert an expression to a plan, using
+            // the rules that the render thinks are appropriate.
+            //
+            // The gist is that we'll want to extract as much map/filter/project as we can
+            // from each expression, attempt to push this down in to the remaining operator,
+            // and
+
+            // Extract a maximally large MapFilterProject from `expr`.
+            // We will then try and push this in to the resulting expression.
+            let (mut mfp, expr) = MapFilterProject::extract_from_expression(expr);
+            // We attempt to plan what we have remaining, in the context of `mfp`.
+            // We may not be able to do this, and must wrap some operators with a `Mfp` stage.
+            let (mfp, plan) = match expr {
+                // These operators should have been extracted from the expression.
+                MirRelationExpr::Map { .. } => {
+                    panic!("This operator should have been extracted");
+                }
+                MirRelationExpr::Filter { .. } => {
+                    panic!("This operator should have been extracted");
+                }
+                MirRelationExpr::Project { .. } => {
+                    panic!("This operator should have been extracted");
+                }
+                // These operators may not have been extracted, and need to result in a `Plan`.
+                MirRelationExpr::Constant { rows, typ: _ } => {
+                    use timely::progress::Timestamp;
+                    let plan = Plan::Constant {
+                        rows: rows.clone().map(|rows| {
+                            rows.into_iter()
+                                .map(|(row, diff)| (row, repr::Timestamp::minimum(), diff))
+                                .collect()
+                        }),
+                    };
+                    (Some(mfp), plan)
+                }
+                MirRelationExpr::Get { id, typ: _ } => (
+                    None,
+                    Plan::Get {
+                        id: id.clone(),
+                        mfp,
+                    },
+                ),
+                MirRelationExpr::Let { id, value, body } => {
+                    // It would be unfortunate to have a non-trivial `mfp` here, as we hope
+                    // that they would be pushed down. I am not sure if we should take the
+                    // initiative to push down the `mfp` ourselves.
```

Comment:
Let's discuss for follow-up work. This was very specifically intended as turning an expression into a dataflow graph, rather than performing optimizations as it goes. For example, I could imagine wanting to let the optimization layer determine whether pushing down a filter makes sense (as it probably "owns" cost-based optimization).